delta-lake/src/main/java/org/apache/iceberg/delta/BaseSnapshotDeltaLakeTableAction.java

```diff
@@ -0,0 +1,370 @@
+      if (action instanceof AddFile) {
+        filesToAdd.add(dataFile);
+      } else if (action instanceof RemoveFile) {
+        filesToRemove.add(dataFile);
+      } else {
+        throw new ValidationException(
+            "The action %s's is unsupported", action.getClass().getSimpleName());
+      }
+    }
+
+    if (filesToAdd.size() > 0 && filesToRemove.size() > 0) {
+      // OverwriteFiles case
+      OverwriteFiles overwriteFiles = transaction.newOverwrite();
+      filesToAdd.forEach(overwriteFiles::addFile);
+      filesToRemove.forEach(overwriteFiles::deleteFile);
+      overwriteFiles.commit();
+    } else if (filesToAdd.size() > 0) {
+      // AppendFiles case
+      AppendFiles appendFiles = transaction.newAppend();
+      filesToAdd.forEach(appendFiles::appendFile);
+      appendFiles.commit();
+    } else if (filesToRemove.size() > 0) {
+      // DeleteFiles case
+      DeleteFiles deleteFiles = transaction.newDelete();
+      filesToRemove.forEach(deleteFiles::deleteFile);
+      deleteFiles.commit();
+    }
+  }
+
+  private DataFile buildDataFileFromAction(Action action, Table table) {
+    PartitionSpec spec = table.spec();
+    String path;
+    Map<String, String> partitionValues;
+
+    if (action instanceof AddFile) {
+      AddFile addFile = (AddFile) action;
+      path = addFile.getPath();
+      partitionValues = addFile.getPartitionValues();
+    } else if (action instanceof RemoveFile) {
+      RemoveFile removeFile = (RemoveFile) action;
+      path = removeFile.getPath();
+      partitionValues = removeFile.getPartitionValues();
+    } else {
+      throw new ValidationException(
+          "Unexpected action type for Delta Lake: %s", action.getClass().getSimpleName());
+    }
+
+    String fullFilePath = getFullFilePath(path, deltaLog.getPath().toString());
+    // For unpartitioned table, the partitionValues should be an empty map rather than null
+    Preconditions.checkArgument(
+        partitionValues != null,
+        String.format("File %s does not specify a partitionValues", fullFilePath));
+
+    FileFormat format = determineFileFormatFromPath(fullFilePath);
+    InputFile file = deltaLakeFileIO.newInputFile(fullFilePath);
+    MetricsConfig metricsConfig = MetricsConfig.forTable(table);
+    String nameMappingString = table.properties().get(TableProperties.DEFAULT_NAME_MAPPING);
+    NameMapping nameMapping =
+        nameMappingString != null ? NameMappingParser.fromJson(nameMappingString) : null;
+    Metrics metrics = getMetricsForFile(file, format, metricsConfig, nameMapping);
+
+    String partition =
+        spec.fields().stream()
+            .map(PartitionField::name)
+            .map(name -> String.format("%s=%s", name, partitionValues.get(name)))
+            .collect(Collectors.joining("/"));
+
+    return DataFiles.builder(spec)
+        .withPath(fullFilePath)
+        .withFormat(format)
+        .withFileSizeInBytes(file.getLength())
```

Comment:
we can save the getLength call incase both RemovedFile and AddFile contains size ?